twitter/src/main/java/com/hazelcast/jet/contrib/twitter/TwitterSources.java

```diff
@@ -0,0 +1,204 @@
+     * StreamSourceStage<String> srcStage = p.readFrom(timestampedStreamSource);
+     * }</pre>
+     *
+     * @param name             a descriptive name of this source.
+     * @param endpointSupplier Supplier that supplies a Twitter StreamingEndpoint to connect to source.
+     * @param credentials      a Twitter OAuth1 credentials that consists "consumerKey",
+     *                         "consumerSecret", "token", "tokenSecret" keys.
+     * @param host             a Twitter host URL to connect.
+     *                         These hosts are defined in {@link com.twitter.hbc.core.Constants}.
+     * @return a timestamped stream source to use in {@link com.hazelcast.jet.pipeline.Pipeline#readFrom}
+     */
+    @Nonnull
+    public static StreamSource<String> timestampedStream(@Nonnull String name,
+                                                         @Nonnull SupplierEx<? extends StreamingEndpoint> endpointSupplier,
+                                                         @Nonnull Properties credentials,
+                                                         @Nonnull String host) {
+        return SourceBuilder.timestampedStream(name,
+                ctx -> new TwitterSourceContext(endpointSupplier, credentials, host))
+                .fillBufferFn(TwitterSourceContext::fillTimestampedBuffer)
+                .destroyFn(TwitterSourceContext::close)
+                .build();
+    }
+
+    /**
+     * A source context of Twitter
+     */
+    private static final class TwitterSourceContext {
+
+        private static final int QUEUE_CAPACITY = 1000;
+        private static final int MAX_FILL_ELEMENTS = 250;
+
+        private final BlockingQueue<String> queue = new LinkedBlockingQueue<>(QUEUE_CAPACITY);
+        private final ArrayList<String> buffer = new ArrayList<>(MAX_FILL_ELEMENTS);
+        private final BasicClient client;
+
+        /**
+         * @param endpointSupplier Supplier that supplies a Twitter StreamingEndpoint to connect to source.
+         * @param credentials      a Twitter OAuth1 credentials that consists "consumerKey",
+         *                         "consumerSecret", "token", "tokenSecret" keys.
+         * @param host             a Twitter host URL to connect.
+         *                         These hosts are defined in {@link com.twitter.hbc.core.Constants}.
+         */
+        private TwitterSourceContext(
+                @Nonnull SupplierEx<? extends StreamingEndpoint> endpointSupplier,
+                @Nonnull Properties credentials,
+                @Nonnull String host
+        ) {
+
+            String consumerKey = credentials.getProperty("consumerKey");
+            String consumerSecret = credentials.getProperty("consumerSecret");
+            String token = credentials.getProperty("token");
+            String tokenSecret = credentials.getProperty("tokenSecret");
+
+            if (isMissing(consumerKey) || isMissing(consumerSecret) || isMissing(token) || isMissing(tokenSecret)) {
```

Comment:
I think it is better to extract this expression to a method with a descriptive name

Reply:
I extract it to the checkTwitterCredentials method.